src/util.cpp

```diff
@@ -943,18 +943,40 @@ void ArgsManager::ReadConfigStream(std::istream& stream)
     }
 }
 
-void ArgsManager::ReadConfigFile(const std::string& confPath)
+void ArgsManager::ReadConfigFiles()
 {
     {
         LOCK(cs_args);
         m_config_args.clear();
+
+        // we do not allow -includeconf from command line, so we clear it here
+        auto it = m_override_args.find("-includeconf");
+        if (it != m_override_args.end()) m_override_args.erase(it);
     }
 
+    const std::string confPath = GetArg("-conf", BITCOIN_CONF_FILENAME);
     fs::ifstream stream(GetConfigFile(confPath));
 
     // ok to not have a config file
     if (stream.good()) {
         ReadConfigStream(stream);
+        std::vector<std::string> includeconf(GetArgs("-includeconf"));
+        {
+            std::vector<std::string> includeconf_net(GetArgs(std::string("-") + GetChainName() + ".includeconf"));
```

Comment:
I was going to say that this isn't required because `ArgsManager::GetArgs()` already checks the network-specific config. However, we haven't yet set `m_network` (that only happens in `SelectParams()`) so you're right and you do need this manual check.

I don't know if there's a better way to avoid doing this manually. Perhaps @ajtowns can suggest something? At the least, I think you should add a comment like:

```
        // We haven't set m_network yet (that happens in SelectParms()), so manually check
        // for network.includeconf args.
```
You can also remove the block/braces.

Reply:
Good point on adding comment. I added braces to throw `includeconf_net` out of scope as it was no longer used, and this makes that obvious.

Reply:
I think this is subtly "wrong" fwiw: if you have `[main] includeconf=foo.conf` in bitcoin.conf, and then set `testnet=1` in foo.conf, you'll have foo.conf included, and be running testnet not mainnet, so you "shouldn't" have included foo.conf (and any other settings from the [main] section won't have been used). That might be too subtle to worry about though.

It's perhaps less subtly wrong in that if you said `includeconf=force-testnet.conf [testnet] includeconf=testnet-params.conf`, with force-testnet.conf setting `testnet=1`, you would not end up including testnet-params.conf, whereas if you set `-testnet` on the command line, you would.

The easiest approach might be to not allow `[regtest] includeconf=rt.conf` at all, but that would probably be annoying (particularly with custom chains).

A thought I had was that maybe `[regtest] includeconf=rt.conf` should treat all the options in rt.conf as being network-specific, so if rt.conf says `connect=10.10.10.10 [main] connect=10.10.10.20 [regtest] connect=10.10.10.30`, that would get interpreted as if it was `[regtest] connect=10.10.10.10 connect=10.10.10.30`. That would avoid the inconsistency above, in that it would turn `[main] includeconf=foo.conf` into essentially `[main] testnet=1` which would just be ignored.

I think it might be better to do it as four phases though: (1) read "-conf", (2) read the includeconfs that aren't under a section, (3) work out the chain based on everything noted so far, and (4) read the "-chain.includeconf" options from -conf (but not letting that change the chain).

Reply:
I think that these are good observations, but that we don't necessarily need to police such pathological edge-cases. The special casing for `[regtest] testnet=1` etc in https://github.com/bitcoin/bitcoin/pull/11862/commits/005ad266491f43d7a9bfd959396037416cb32a55 was useful because it's easy enough to have a config file like:

```
option1=0
option2=1
[regtest]
option3=1
```

and then append `testnet=1` to the end, without realising that you're actually adding `regtest.testnet=1`. (in fact, this PR does something similar here: https://github.com/bitcoin/bitcoin/pull/10267/files#diff-a3677de1e28a0e9b37d66c11cc9a760bR31 by appending an argument that becomes a network-specific argument). This kind of config error is easy to make, ambiguous, and ought to be policed.

The config that you're describing here:

> if you said includeconf=force-testnet.conf [testnet] includeconf=testnet-params.conf, with force-testnet.conf setting testnet=1, you would not end up including testnet-params.conf, whereas if you set -testnet on the command line, you would.

seems much more contorted. I don't think a user should reasonably expect that setting the network in an `includeconf` file should change how the network-specific `includeconf`s in the base config file are interpreted. That behaviour also feels a bit too multi-leveled in that we're reading a specific `includeconf` file based on the content of a different `includeconf` file (I argued against such behaviour here: https://github.com/bitcoin/bitcoin/pull/10267#issuecomment-301810263 because I think it leads to an overly complex model).

Reply:
@kallewoof 

>  I added braces to throw includeconf_net out of scope

I don't know of any other places in the codebase that add blocks to throw local variables out of scope. Obviously functionally the same, but I'd personally have a preference to not add code blocks like this, since it makes the code slightly less clear.

Reply:
Agree with @jnewbery, there is no reason to do this. At first sight I thought there was a lock somewhere.